### PR TITLE
Implementation of held/unheld functions for state pkg

### PR DESCRIPTION
**Pull request by vzhestkov**
### What does this PR do?

Implements `held`/`unheld` functions for state `pkg` module.
Examples:
```
sysstat:
  pkg.held: []
```
```
held-pkgs:
  pkg.held:
    - pkgs:
      - sysstat
      - vnstat: 1.17-bp153.1.18
```
```
held-pkgs-clear:
  pkg.held:
    - pkgs: []
    - replace: True
```
```
unheld-pkgs:
  pkg.unheld:
    - pkgs:
      - sysstat
      - vnstat: 1.17-bp153.1.18
      - test-package
```
```
unheld-all:
  pkg.unheld:
    - all: True
```

Implements `list_holds` function for `zypperpkg` module. The output is similar to `yumpkg.list_holds`. The only difference is that `zypperpkg` calculates the current locked packages from `list_locks` as the yum and zypper have different way of managing locks.

Improves `hold` function for `zypperpkg` to handle package version if specified.

### CODE REVIEW

**Review by Ch3LL — CHANGES_REQUESTED** [review]
this will require a changelog and test coverage

File: salt/modules/zypperpkg.py [hunk 1]
```diff
@@ -2106,28 +2179,42 @@ def unhold(name=None, pkgs=None, **kwargs):
 
     targets = []
     if pkgs:
-        for pkg in salt.utils.data.repack_dictlist(pkgs):
```
Comment (Ch3LL) [inline salt/modules/zypperpkg.py:2109]: Why are you removing this?

File: salt/modules/zypperpkg.py [hunk 2]
```diff
@@ -2205,18 +2295,20 @@ def hold(name=None, pkgs=None, **kwargs):
...
```
Comment (Ch3LL) [inline salt/modules/zypperpkg.py:2208]: here as well. why removing?

File: salt/modules/zypperpkg.py [hunk 3]
```diff
@@ -2558,6 +2650,9 @@ def search(criteria, refresh=False, **kwargs):
     details (bool)
         Show version and repository
 
+    ignore_no_matching_item (bool)
```
Comment (Ch3LL) [inline salt/modules/zypperpkg.py:2653]: I'm not seeing this kwarg being used in this function.

**Review by Ch3LL — CHANGES_REQUESTED** [review]

File: salt/modules/zypperpkg.py [hunk 4]
```diff
@@ -2007,6 +2007,76 @@ def purge(name=None, pkgs=None, root=None, **kwargs):  # pylint: disable=unused-
...
+            continue
+        try:
+            found_pkgs = search(
+                solv_name,
+                root=root,
+                match=None if "*" in solv_name else "exact",
+                case_sensitive=(lock.get("case_sensitive", "on") == "on"),
+                installed_only=True,
+                details=True,
+                all_versions=True,
```
[the remark this answers is not in the document]
Comment (vzhestkov) [reply salt/modules/zypperpkg.py:2050]: Just realized that `all_version` left there as `info_installed` was used before, but then I've changed it to `search`.
`ignore_no_matching_item` is a side effect of implementing the fix which is actually should be fixed with this PR:https://github.com/saltstack/salt/pull/58552 I reverted the change required to fix this issue, but not fully. Will fix it. Thanks

File: salt/modules/zypperpkg.py (hunk 1 above)
Comment (vzhestkov) [reply salt/modules/zypperpkg.py:2109]: This for loop removes the versions from the `pkgs` list, but in `zypper` the package could be locked with the specified version

File: salt/modules/zypperpkg.py (hunk 2 above)
Comment (vzhestkov) [reply salt/modules/zypperpkg.py:2208]: same as above

File: salt/modules/zypperpkg.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (vzhestkov) [reply salt/modules/zypperpkg.py]: Fixed with https://github.com/saltstack/salt/pull/60432/commits/a685243947cfcd383b60101ab5c4f2afd7f408e6

File: salt/modules/zypperpkg.py [hunk 5]
```diff
@@ -2018,6 +2018,73 @@ def purge(name=None, pkgs=None, root=None, **kwargs):  # pylint: disable=unused-
     return _uninstall(name=name, pkgs=pkgs, root=root)
 
 
+def list_holds(pattern=None, full=True, root=True, **kwargs):
+    """
+    List information on locked packages.
+
```
Comment (twangboy) [inline salt/modules/zypperpkg.py:2024]: This function needs a `.. versionadded:: Phosphorus` line in the docs

File: salt/states/pkg.py [hunk 6]
```diff
@@ -3629,3 +3629,313 @@ def mod_beacon(name, **kwargs):
             ),
             "result": False,
         }
+
+
+def held(name, version=None, pkgs=None, replace=False, **kwargs):
+    """
+    Set package in 'hold' state, meaning it will not be changed.
```
Comment (twangboy) [inline salt/states/pkg.py:3636]: This new function also needs a `.. versionadded:: Phosphorus` docs marker

File: salt/states/pkg.py [hunk 7]
```diff
@@ -3629,3 +3629,313 @@ def mod_beacon(name, **kwargs):
...
+    ret["comment"] = "\n".join(comments)
+    if not (ret["changes"] or ret["comment"]):
+        ret["comment"] = "No changes made"
+
+    return ret
+
+
+def unheld(name, version=None, pkgs=None, all=False, **kwargs):
+    """
+    Unset package from 'hold' state, to allow operations with the package.
```
Comment (twangboy) [inline salt/states/pkg.py:3813]: `.. versionadded:: Phosphorus`

**Review by twangboy — CHANGES_REQUESTED** [review]

**Comment by garethgreenaway** [comment]
@vzhestkov Some merge conflicts on this one and some requested changes.

File: salt/modules/zypperpkg.py (hunk 5 above)
Comment (twangboy) [reply salt/modules/zypperpkg.py:2022]: If you want this to go in for Phosphorus, we need to have versionadded. But it's 3005 now so:
```
.. versionadded:: 3005
```

File: salt/states/pkg.py (hunk 6 above)
Comment (twangboy) [reply salt/states/pkg.py:3649]: If you want this to go in for Phosphorus, we need to have versionadded. But it's 3005 now so:
```
.. versionadded:: 3005
```

File: salt/states/pkg.py (hunk 7 above)
Comment (twangboy) [reply salt/states/pkg.py:3826]: If you want this to go in for Phosphorus, we need to have versionadded. But it's 3005 now so:
```
.. versionadded:: 3005
```

**Review by twangboy — CHANGES_REQUESTED** [review]

**Review by twangboy — APPROVED** [review]

**Comment by vzhestkov** [comment]
@twangboy thanks a lot for taking care of it

**Review by Ch3LL — APPROVED** [review]

**Pull request merged**